tests/test_power_profiles_daemon.py

```diff
@@ -0,0 +1,148 @@
+__copyright__ = '(c) 2021 Red Hat Inc.'
+
+import fcntl
+import os
+import shutil
+import subprocess
+import sys
+import unittest
+import time
+
+import dbus
+import dbus.mainloop.glib
+
+import dbusmock
+
+dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
+
+have_powerprofilesctl = shutil.which('powerprofilesctl')
+
+
+@unittest.skipUnless(have_powerprofilesctl, 'powerprofilesctl not installed')
+class TestPowerProfilesDaemon(dbusmock.DBusTestCase):
+    '''Test mocking power-profiles-daemon'''
+    @classmethod
+    def setUpClass(cls):
+        cls.start_system_bus()
+        cls.dbus_con = cls.get_dbus(True)
+
+    def setUp(self):
+        (self.p_mock, self.obj_ppd) = self.spawn_server_template(
+            'power_profiles_daemon', {}, stdout=subprocess.PIPE)
+        # set log to nonblocking
+        flags = fcntl.fcntl(self.p_mock.stdout, fcntl.F_GETFL)
+        fcntl.fcntl(self.p_mock.stdout, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+        self.dbusmock = dbus.Interface(self.obj_ppd, dbusmock.MOCK_IFACE)
+
+    def tearDown(self):
+        self.p_mock.stdout.close()
+        self.p_mock.terminate()
+        self.p_mock.wait()
+
+    def test_list_profiles(self):
+        '''List Profiles and check active profile'''
+
+        out = subprocess.check_output(['powerprofilesctl'],
+                                      universal_newlines=True)
+
+        self.assertIn('performance:\n', out)
+        self.assertIn('\n* balanced:\n', out)
+
+    def test_change_profile(self):
+        '''Change ActiveProfile'''
+
+        subprocess.check_output(['powerprofilesctl', 'set', 'performance'],
+                                universal_newlines=True)
+        out = subprocess.check_output(['powerprofilesctl', 'get'],
+                                      universal_newlines=True)
+        self.assertEqual(out, 'performance\n')
+
+    def run_powerprofilesctl_list_holds(self):  # pylint: disable=no-self-use
+        return subprocess.check_output(['powerprofilesctl', 'list-holds'],
+                                       universal_newlines=True)
+
+    def test_list_holds(self):
+        '''Test holds'''
+
+        # No holds
+        out = self.run_powerprofilesctl_list_holds()
+        self.assertEqual(out, '')
+
+        # 1 hold
+        cmd = subprocess.Popen(['powerprofilesctl', 'launch', '-p',
+                                'power-saver', '-r', 'g-s-d mock test',
+                                '-i', 'org.gnome.SettingsDaemon.Power',
+                                'sleep', '60'],
+                               stdout=subprocess.PIPE)
+        time.sleep(0.3)
+
+        out = self.run_powerprofilesctl_list_holds()
+        self.assertEqual(out, 'Hold:\n'
+                              '  Profile:         power-saver\n'
+                              '  Application ID:'
+                              '  org.gnome.SettingsDaemon.Power\n'
+                              '  Reason:          g-s-d mock test\n')
```

Comment:
Hint for the future: Multi-line strings with `'''` or `"""` are a bit nicer for this.

Reply:
I forgot about this, but this would actually make the string less readable, as we'd need to nuke the indentation, and rely on implicit linefeeds.